sql/sql_select.cc

```diff
@@ -693,52 +703,115 @@ setup_for_system_time(THD *thd, TABLE_LIST *tables, COND **conds, SELECT_LEX *se
   if (versioned_tables == 0)
     DBUG_RETURN(0);
 
+  /* For prepared statements we create items on statement arena,
+     because they must outlive execution phase for multiple executions. */
+  arena= thd->activate_stmt_arena_if_needed(&backup);
+
+  if (select_lex->saved_conds)
+  {
+    DBUG_ASSERT(thd->stmt_arena->is_sp_execute());
+    *conds= select_lex->saved_conds;
+  }
+  else if (thd->stmt_arena->is_sp_execute())
+  {
+    if (thd->stmt_arena->is_stmt_execute())
+      *conds= 0;
+    else if (*conds)
+      select_lex->saved_conds= (*conds)->copy_andor_structure(thd);
+  }
+
   for (table= tables; table; table= table->next_local)
   {
-    if (table->table && table->table->versioned_by_sql())
+    if (table->table && table->table->versioned())
     {
       Field *fstart= table->table->vers_start_field();
       Field *fend= table->table->vers_end_field();
-      Item *istart= new (thd->mem_root) Item_field(thd, fstart);
-      Item *iend= new (thd->mem_root) Item_field(thd, fend);
-      Item *cond1= 0, *cond2= 0, *curr = 0;
+
+      DBUG_ASSERT(select_lex->parent_lex);
+      Name_resolution_context *context= select_lex->parent_lex->current_context();
+      DBUG_ASSERT(context);
+
+      Item *row_start= new (thd->mem_root) Item_field(thd, context, fstart);
+      Item *row_end= new (thd->mem_root) Item_field(thd, context, fend);
+      Item *row_end2= row_end;
+
+      if (!table->table->versioned_by_sql())
+      {
+        DBUG_ASSERT(table->table->s && table->table->s->db_plugin);
+        row_start= new (thd->mem_root) Item_func_vtq_ts(
+          thd,
+          row_start,
+          VTQ_COMMIT_TS,
```

Comment:
In this block we create 2 functions: one for getting begin_ts and one for commit_ts. One Item_func_vtq_ts instance receives row_start and should receive VTQ_BEGIN_TS, and the second one receives row_end and should reeive VTQ_COMMIT_TS. This looks like a copy/paste error.